msal/token_cache.py

```diff
@@ -136,20 +143,20 @@ def search(self, credential_type, target=None, query=None):  # O(n) generator
             preferred_result = self._get_access_token(
                 query["home_account_id"], query["environment"],
                 query["client_id"], query["realm"], target)
-            if preferred_result:
+            if preferred_result and self._is_matching(
+                preferred_result, query, target_set,
+            ):
                 yield preferred_result
 
-        target_set = set(target)
         with self._lock:
             # Since the target inside token cache key is (per schema) unsorted,
             # there is no point to attempt an O(1) key-value search here.
             # So we always do an O(n) in-memory search.
             for entry in self._cache.get(credential_type, {}).values():
-                if is_subdict_of(query or {}, entry) and (
-                        target_set <= set(entry.get("target", "").split())
-                        if target else True):
-                    if entry != preferred_result:  # Avoid yielding the same entry twice
-                        yield entry
```

Comment:
Embedded `if` statements can be combined with `and`.

Reply:
Originally it contained only the first `if`. Later when we added the second `if`, we did it that way so that the PR at that time was easier to review (i.e. less "red" background in a PR).

Now that after this PR's refactoring, we indeed combine them into one short and readable `if`.